src/Microsoft.DncEng.SecretManager/SecretTypes/AzureStorageBlobSas.cs

```diff
@@ -0,0 +1,39 @@
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.WindowsAzure.Storage.Blob;
+
+namespace Microsoft.DncEng.SecretManager.SecretTypes
+{
+    [Name("azure-storage-blob-sas-uri")]
+    public class AzureStorageBlobSas : AzureStorageSasSecretType
+    {
+        private readonly string _containerName;
+        private readonly string _blobName;
+        private readonly string _permissions;
+        public AzureStorageBlobSas(IReadOnlyDictionary<string, string> parameters) : base(parameters)
+        {
+            ReadRequiredParameter("blob", ref _blobName);
+            ReadRequiredParameter("container", ref _containerName);
+            ReadRequiredParameter("permissions", ref _permissions);
+        }
+
+        protected override async Task<SecretData> RotateValue(RotationContext context, CancellationToken cancellationToken)
+        {
+            var now = DateTimeOffset.UtcNow;
```

Comment:
This should probably be an ISystemClock so we can add testing here if we want. :-) I think we inject one in the Program defaults?

Reply:
I can do that, but I'm not super convinced that unit testing for these types is worth the effort. Each one is a thin wrapper around the underlying azure apis, so we would need to mock those somehow to unit test this; and the types themselves don't have many if any branches, so the value is low. Integration tests can cover these. I will change this to use the ISystemClock though so we can more easily add unit tests in the future if they are deemed valuable.

Reply:
Oooh, these types aren't DI'd... it'll be a pain in the butt to do that.  It's probably not worth it. :-)  You'd end up having to split them in annoyingly complicated ways or do odd reflection.

Reply:
You'd probably want a "StorageBlobSasParameters" object that _wasn't_ DI'd, and had no actual code in it, and then an IStorageBlobSasRotator.Rotate(Parameters) method that did the actual work.

Then you'd have a singleton rotator, rather than having every data object be able to rotate itself.

Reply:
they are DIed actually, just I have to pass them some parameters too, so it uses the ActivatorUtilities.CreateInstance thingy.

Reply:
Note that those aren't verifyable (since the verifier has no way to know way to know which parameters are "real" and which aren't), so that might let DI bugs slip in, which are super annoying to deal with. :-)

Reply:
It feels like mixing dependencies and data parameters in the same constructor is a sort of fuzzy design.  It sort of means the class is doing two things... it's both representing data, but also embodying behavioral logic in the same class.